Support Cirrus Logic CS42L42 headphone codec

brentlu opened the pull request:
It's a series to:
1. Add a machine driver, sof_cs42l42, to support Cirrus Logic CS42L42 + Maxim MAX98357A on GLK devices.
2. Move MAX98357A support to a common module, sof_maxim_common
3. Code refactory to sof_rt5682 and sof_cs42l42 to use sof_maxim_common to support Maxim MAX98357A

brentlu commented:
The topology is here: https://github.com/thesofproject/sof/pull/4121

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,551 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright(c) 2019-2020 Intel Corporation.
```
2021

brentlu replied:
fixed

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,551 @@
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOLUMEUP);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEDOWN);
+	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOICECOMMAND);
```
is the mapping correct? It seems that DA7219 had a quirk compared to the other codecs, the regular sequence is

````
	snd_jack_set_key(jack->jack, SND_JACK_BTN_0, KEY_PLAYPAUSE);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_1, KEY_VOICECOMMAND);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_2, KEY_VOLUMEUP);
	snd_jack_set_key(jack->jack, SND_JACK_BTN_3, KEY_VOLUMEDOWN);
````

brentlu replied:
Yes, ODM validates the pause/volume up/volume down buttons.

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,551 @@
+	 * ignore_pmdown_time to power down rt5682 immediately and
+	 * avoid the noise.
+	 * It can be removed once we can control MCLK by driver.
+	 */
```
this looks like copy-paste, is this .ignore_pmdow_time necessary?

brentlu replied:
According to ODM test result, there is no pop/static noise after removing this option. So I remove it.

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,551 @@
+			 SOF_CS42L42_NUM_HDMIDEV_SHIFT;
+		/* default number of HDMI DAI's */
+		if (!hdmi_num)
+			hdmi_num = 3;
```
this looks like copy-paste, we don't have any other platform with cs32l42.

brentlu replied:
Suggest keeping it since we are going to support ADL platform in the future.

plbossart replied:
ok

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,551 @@
+/* Module information */
+MODULE_DESCRIPTION("SOF Audio Machine driver");
+MODULE_AUTHOR("Brent Lu <brent.lu@intel.com>");
+MODULE_LICENSE("GPL v2");
```
"GPL"

also rename the 'SOF Audio Machine driver', it's too vague, you need a reference to cs42l42

brentlu replied:
fixed

plbossart commented on sound/soc/intel/boards/Kconfig:
```diff
@@ -479,6 +479,23 @@ config SND_SOC_INTEL_SOF_RT5682_MACH
+	depends on I2C && ACPI
+	depends on ((SND_HDA_CODEC_HDMI && SND_SOC_SOF_HDA_AUDIO_CODEC) &&\
+		    (MFD_INTEL_LPSS || COMPILE_TEST)) ||\
+		   (SND_SOC_SOF_BAYTRAIL && (X86_INTEL_LPSS || COMPILE_TEST))
```
we can simplify this, we don't have a case with BAYTRAIL.
@kv2019i can you double-check the HDaudio parts, I can't recall what's necessary.

brentlu replied:
removed

plbossart commented on sound/soc/intel/boards/sof_rt5682.c:
```diff
@@ -451,20 +451,26 @@ static int sof_card_late_probe(struct snd_soc_card *card)
 static const struct snd_kcontrol_new sof_controls[] = {
```
comment on commit message: I think you meant 'refactor'

brentlu replied:
fixed

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -549,3 +521,4 @@ MODULE_LICENSE("GPL v2");
 MODULE_ALIAS("platform:sof_cs42l42");
 MODULE_ALIAS("platform:glk_cs4242_max98357a");
 MODULE_IMPORT_NS(SND_SOC_INTEL_HDA_DSP_COMMON);
+MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_MAXIM_COMMON);
```
you can move the maxim common part first, as a preparation, and then squash all the cs42l42 changes. No need to show the changes in separate patches, you can just add the new machine driver in one patch, i'll be simpler to review.

brentlu replied:
Rebase the commits to merge cs42l42 changes.

lyakh commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,514 @@
+	return ret;
+}
+
+static struct snd_soc_ops sof_cs42l42_ops = {
```
this can be `const`

brentlu replied:
fixed

lyakh commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,514 @@
+		return ret;
+	}
+
+	jack = &ctx->headset_jack;
```
nitpick: if you do this right at the top of the function, maybe just in the initialisation, you can also use `jack` in `snd_soc_card_jack_new()` call above

brentlu replied:
done

lyakh commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,514 @@
+
+		if (sof_cs42l42_quirk & SOF_MAX98357A_SPEAKER_AMP_PRESENT) {
+			max_98357a_dai_link(&links[id]);
+		} else
```
braces shouldn't be used

brentlu replied:
fixed, thanks for the review~

lyakh reviewed (commented):
some minor cosmetic improvement suggestions

plbossart reviewed (approved):
thanks @brentlu, nice work.

plbossart commented:
I had to manually approve @brentlu so let's wait until we see the GitHub Actions results + another approved to merge

plbossart reviewed (changes requested):
I have to take my approval back after looking at the topology.
It's simpler if we have first the SSP links added, then dmic and HDMI last. The topology changes [1] can be simplified a lot by using the same order as for DA7219.

[1] https://github.com/thesofproject/sof/pull/4121/commits/ef0ebafad22e2cf46dc6bc40396d75d9590dd2f4#diff-ef984278da4b010127ccd588daa6d53c55bff7a852c2ef42f7b39a82d4bd3951R195

brentlu commented:
Adjust the DAI Link sequence in sof_card_dai_links_create() function to follow the order: SPK, HEADPHONE, DMIC, HDMI

plbossart commented:
restarting SOF CI tests since we are missing the configuration for the new machine driver

plbossart commented:
SOFCI TEST

plbossart reviewed (changes requested):
@brentlu I added the missing config and CI reports errors with make W=1 and also checkpatch.pl --strict has quite a few warnings. Nothing critical but let's make this PR shiny for upstream. Thanks!

bardliao commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,516 @@
+
+	/* Configure sysclk for codec */
+	ret = snd_soc_dai_set_sysclk(codec_dai, 0,
+				     clk_freq, SND_SOC_CLOCK_IN);
```
nitpick: How about remove `clk_freq` and just use `3072000` here?

brentlu replied:
we may switch to 2.4 MHz in the future so suggest keeping it.

bardliao commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,516 @@
+		if (err < 0)
+			return err;
+
+		i++;
```
Sorry, I didn't get what is `i` for?

brentlu replied:
seems useless so i remove it. thanks~

bardliao reviewed (commented):
Looks quite good to me, just some minor comments.

brentlu commented:
warnings from checkpatch are fixed

brentlu commented:
False alarm?

-------------------------------------------------------------------
Commit 429ffb620dcd ("ASoC: Intel: add sof-cs42l42 machine driver")
-------------------------------------------------------------------
WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
#74: 
new file mode 100644

total: 0 errors, 1 warnings, 0 checks, 570 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Commit 429ffb620dcd ("ASoC: Intel: add sof-cs42l42 machine driver") has style problems, please review.

brentlu commented:
@plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?

plbossart commented:
> False alarm?
> 
> ## Commit [429ffb6](https://github.com/thesofproject/linux/commit/429ffb620dcdda38bde01b36be030e2fe52dc848) ("ASoC: Intel: add sof-cs42l42 machine driver")
> WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?
> #74:
> new file mode 100644
> 
> total: 0 errors, 1 warnings, 0 checks, 570 lines checked
> 
> NOTE: For some of the reported defects, checkpatch may be able to
> mechanically convert to the typical style using --fix or --fix-inplace.
> 
> Commit [429ffb6](https://github.com/thesofproject/linux/commit/429ffb620dcdda38bde01b36be030e2fe52dc848) ("ASoC: Intel: add sof-cs42l42 machine driver") has style problems, please review.

yes, that's normal and not a problem.

plbossart commented:
> @plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?

let me check with Mark Brown, there's a missing dependency upstream, some previous patches were not applied, and if you send the patches they will not apply directly. thanks!

plbossart commented:
still a make W=1 issue @brentlu . You can fix this in your setup by running 'make sound/soc/codecs/ W=1' (the last / matters). Thanks!

plbossart commented:
> > @plbossart If there is no concern, I'd like to send the patches to alsa-dev due to the tight schedule. Is it ok?
> 
> let me check with Mark Brown, there's a missing dependency upstream, some previous patches were not applied, and if you send the patches they will not apply directly. thanks!

ok, so I need to resend the previous series which was partly missed for some reason, and I'll include your PR on top once all the warnings are removed. Does this work for you @brentlu? Thanks!

brentlu commented:
The warning is from the codec header file which we don't need it. I've removed it from machine driver and send a mail to Cirrus team to fix it.

brentlu commented:
Thanks. I think it's the best option.

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,511 @@
+
+		links[id].platforms = platform_component;
+		links[id].num_platforms = ARRAY_SIZE(platform_component);
+		links[id].nonatomic = true;
```
sorry, forgot this one. This was removed in an earlier commit in other platforms, for consistency please remove the .nonatomic here as well.

brentlu replied:
removed

plbossart commented on sound/soc/intel/boards/sof_cs42l42.c:
```diff
@@ -0,0 +1,511 @@
+	links[id].init = sof_cs42l42_init;
+	links[id].exit = sof_cs42l42_exit;
+	links[id].ops = &sof_cs42l42_ops;
+	links[id].nonatomic = true;
```
remove

brentlu replied:
remove, BTW, how do I tell if the flag is needed or not?

plbossart replied:
this is not needed for BE, it's only needed for FE, and FEs are now handled in the topology. You don't need to worry about it for newer machine drivers.

plbossart replied:
see commit 961cd002aacc9ff027f1eabdcd5b8cd0c35346a6 @brentlu

plbossart reviewed (changes requested):
sorry, one more fix needed, the code was inspired by an earlier version of sof-rt5682 but the .nonatomic property is not needed.

plbossart reviewed (approved):
let's merge, thanks @brentlu

plbossart commented:
I'll use my superpowers to merge and avoid more delays
